substrate/primitives/virtualization/src/native.rs

```diff
@@ -0,0 +1,331 @@
+// This file is part of Substrate.
+
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::{
+	ExecError, InstantiateError, MemoryError, MemoryT, SharedState, SyscallHandler, VirtT,
+};
+use polkavm::{
+	Caller, CallerRef, Config, Engine, ExecutionError, GasMeteringKind, Instance, Linker, Module,
+	ModuleConfig, Reg, StateArgs, Trap,
+};
+use std::{
+	cell::RefCell,
+	mem,
+	rc::{Rc, Weak},
+	sync::OnceLock,
+};
+
+/// This is the single PolkaVM engine we use for everything.
+///
+/// By using a common engine we allow PolkaVM to use caching. This caching is important
+/// to reduce startup costs. This is even the case when instances use different code.
+static ENGINE: OnceLock<Engine> = OnceLock::new();
+
+/// Engine wide configuration.
+fn engine() -> &'static Engine {
+	ENGINE.get_or_init(|| {
+		let config = Config::new();
+		Engine::new(&config).expect("Default config is always valid; qed")
+	})
+}
+
+/// Native implementation of [`VirtT`].
+pub struct Virt {
```

Comment:
can we prefix with Native and Forwarder each Virt struct to make it easier to grep code when searching for one or the other.

Reply:
Modules are the idiomatic way in Rust to namespace types rather than prefixing their names. Why would an exception be justified here?

Reply:
fair, it just find it a bit more convenient here to find what you want when you just rg `virt`